infer/tests/provider.go

```diff
@@ -173,13 +173,134 @@ func (*WiredPlus) WireDependencies(f infer.FieldSelector, a *WiredInputs, s *Wir
 	r.WireDependencies(f, a, &s.WiredOutputs)
 }
 
+// Default values are applied by the provider to facilitate integration testing and to
+// backstop non-compliment SDKs.
+
+// TODO[pulumi-go-provider#98] Remove the ,optional.
+
+type WithDefaults struct{}
+type WithDefaultsOutput struct{ WithDefaultsArgs }
+
+var (
+	_ infer.Annotated = (*WithDefaultsArgs)(nil)
+	_ infer.Annotated = (*NestedDefaults)(nil)
+)
+
+type WithDefaultsArgs struct {
+	// We sanity check with some primitive values, but most of this checking is in
+	// NestedDefaults.
+	String       string                     `pulumi:"s,optional"`
+	IntPtr       *int                       `pulumi:"pi,optional"`
+	Nested       NestedDefaults             `pulumi:"nested,optional"`
+	NestedPtr    *NestedDefaults            `pulumi:"nestedPtr,optional"`
+	ArrNested    []NestedDefaults           `pulumi:"arrNested,optional"`
+	ArrNestedPtr []*NestedDefaults          `pulumi:"arrNestedPtr,optional"`
+	MapNested    map[string]NestedDefaults  `pulumi:"mapNested,optional"`
+	MapNestedPtr map[string]*NestedDefaults `pulumi:"mapNestedPtr,optional"`
+
+	NoDefaultsPtr *NoDefaults `pulumi:"noDefaults,optional"`
+}
+
+// We want to make sure we don't effect structs or maps that don't have default values.
+type NoDefaults struct {
+	String string `pulumi:"s,optional"`
+}
+
+func (w *WithDefaultsArgs) Annotate(a infer.Annotator) {
+	a.SetDefault(&w.String, "one")
+	a.SetDefault(&w.IntPtr, 2)
+}
+
+type NestedDefaults struct {
+	// Direct vars. These don't allow setting zero values.
+	String string  `pulumi:"s,optional"`
+	Float  float64 `pulumi:"f,optional"`
+	Int    int     `pulumi:"i,optional"`
+	Bool   bool    `pulumi:"b,optional"`
+
+	// Indirect vars. These should allow setting zero values.
+	StringPtr *string  `pulumi:"ps,optional"`
+	FloatPtr  *float64 `pulumi:"pf,optional"`
```

Comment:
Curious, are these types permitted here never Output or Input types? The reflective pass does not deal with Output/Input which makes it simple, but I was curious if that's warranted?

Reply:
They are never input or output types. The structs for custom resources deal exclusively in plain values.